### Task 3 : Encoding Strategy Comparison 
This notebook takes a dataset of airline passenger satisfaction from Kaggle and then applies various encoding techniques to each categorical column. Then train a logistic regression model in each encoded dataset. Finally, measure and compare accuracy, training time, and feature count for each strategy.


In [111]:
import pandas as pd
from sklearn.model_selection import train_test_split 
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, OrdinalEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import time

#### Read csv file

In [112]:
train_df = pd.read_csv("airplane_passenger_dataset/train.csv")
test_df = pd.read_csv("airplane_passenger_dataset/test.csv")

# Remove unnecessary index columns
train_df = train_df.drop(columns=["Unnamed: 0", "id"])
test_df = test_df.drop(columns=["Unnamed: 0", "id"])

# Combine both datasets
df = pd.concat(
    [train_df, test_df],
    ignore_index=True
)

df.head()

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,1,...,5,4,3,4,4,5,5,25,18.0,neutral or dissatisfied
1,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,3,...,1,1,5,3,1,4,1,1,6.0,neutral or dissatisfied
2,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,2,...,5,4,3,4,4,4,5,0,0.0,satisfied
3,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,5,...,2,2,5,3,1,4,2,11,9.0,neutral or dissatisfied
4,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,3,...,3,3,4,4,3,3,3,0,0.0,satisfied


#### Finding categorical columns

In [113]:
df.dtypes

Gender                                   str
Customer Type                            str
Age                                    int64
Type of Travel                           str
Class                                    str
Flight Distance                        int64
Inflight wifi service                  int64
Departure/Arrival time convenient      int64
Ease of Online booking                 int64
Gate location                          int64
Food and drink                         int64
Online boarding                        int64
Seat comfort                           int64
Inflight entertainment                 int64
On-board service                       int64
Leg room service                       int64
Baggage handling                       int64
Checkin service                        int64
Inflight service                       int64
Cleanliness                            int64
Departure Delay in Minutes             int64
Arrival Delay in Minutes             float64
satisfacti

In [114]:
# Check missing values in the dataset
df.isnull().sum()

Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Inflight wifi service                  0
Departure/Arrival time convenient      0
Ease of Online booking                 0
Gate location                          0
Food and drink                         0
Online boarding                        0
Seat comfort                           0
Inflight entertainment                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Inflight service                       0
Cleanliness                            0
Departure Delay in Minutes             0
Arrival Delay in Minutes             393
satisfaction                           0
dtype: int64

In [115]:
missing_percentage = (
    df["Arrival Delay in Minutes"].isnull().sum() / len(df)
) * 100

print(missing_percentage)

0.3025870033877425


In [116]:
# Fill numerical missing values with median
df["Arrival Delay in Minutes"] = df["Arrival Delay in Minutes"].fillna(
    df["Arrival Delay in Minutes"].median()
)

# Check again
print(df.isnull().sum())

Gender                               0
Customer Type                        0
Age                                  0
Type of Travel                       0
Class                                0
Flight Distance                      0
Inflight wifi service                0
Departure/Arrival time convenient    0
Ease of Online booking               0
Gate location                        0
Food and drink                       0
Online boarding                      0
Seat comfort                         0
Inflight entertainment               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Inflight service                     0
Cleanliness                          0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
satisfaction                         0
dtype: int64


In [ ]:
# Selecting input features and target column
X = df.drop("satisfaction", axis=1)
y = df["satisfaction"]

In [145]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# feature count(column count) before any encoding
X_train.shape[1]

22

### Strategy 1: Label Encoding

In [119]:
# Find categorical columns
categorical_columns = X_train.select_dtypes(include="str").columns

In [140]:
# Copy data so original data is not changed
X_train_label = X_train.copy()
X_test_label = X_test.copy()

22

In [146]:
# Apply Label Encoding
for col in categorical_columns:
    encoder = LabelEncoder()

    # Learn categories from training data
    encoder.fit(X_train_label[col])

    # Convert train and test using same encoder
    X_train_label[col] = encoder.transform(X_train_label[col])
    X_test_label[col] = encoder.transform(X_test_label[col])

print("Label Encoding completed;")
print(X_train_label.head())

Label Encoding completed;
       Gender  Customer Type  Age  Type of Travel  Class  Flight Distance  \
21413       0              0   64               1      0              965   
7284        1              1   26               0      0              592   
86080       0              0    8               1      1              993   
59842       1              0   41               0      0             2223   
69486       0              1   19               0      1             1235   

       Inflight wifi service  Departure/Arrival time convenient  \
21413                      3                                  5   
7284                       2                                  1   
86080                      3                                  4   
59842                      4                                  4   
69486                      4                                  4   

       Ease of Online booking  Gate location  ...  Seat comfort  \
21413                       3            

In [150]:
# feature count(column count) after Label Encoding
X_train_label.shape[1]

22

In [147]:
# Summarizes the statistical distribution of data
X_train_label.describe()

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,...,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
count,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.00000,103904.000000,103904.000000,...,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000
mean,0.493071,0.183987,39.402189,0.309738,0.594722,1188.271193,2.725843,3.05969,2.755043,2.979269,...,3.440147,3.357416,3.383421,3.351305,3.633267,3.305282,3.645095,3.287063,14.743850,15.054810
std,0.499954,0.387476,15.123232,0.462388,0.621407,995.924038,1.329907,1.52637,1.402951,1.276298,...,1.319246,1.334211,1.288294,1.315670,1.179274,1.265602,1.175694,1.311912,38.108621,38.460237
min,0.000000,0.000000,7.000000,0.000000,0.000000,31.000000,0.000000,0.00000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,27.000000,0.000000,0.000000,413.000000,2.000000,2.00000,2.000000,2.000000,...,2.000000,2.000000,2.000000,2.000000,3.000000,3.000000,3.000000,2.000000,0.000000,0.000000
50%,0.000000,0.000000,40.000000,0.000000,1.000000,843.000000,3.000000,3.00000,3.000000,3.000000,...,4.000000,4.000000,4.000000,4.000000,4.000000,3.000000,4.000000,3.000000,0.000000,0.000000
75%,1.000000,0.000000,51.000000,1.000000,1.000000,1739.000000,4.000000,4.00000,4.000000,4.000000,...,5.000000,4.000000,4.000000,4.000000,5.000000,4.000000,5.000000,4.000000,12.000000,13.000000
max,1.000000,1.000000,85.000000,1.000000,2.000000,4983.000000,5.000000,5.00000,5.000000,5.000000,...,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,1305.000000,1280.000000


#### Scale numerical features

In [148]:
# Select numerical columns
numeric_columns = X_train_label.select_dtypes(include=["int64", "float64"] ).columns

# Create scaler
scaler = StandardScaler()

# Learn scaling values from training data
X_train_label[numeric_columns] = scaler.fit_transform( X_train_label[numeric_columns] )


# Apply same scaling to test data
X_test_label[numeric_columns] = scaler.transform( X_test_label[numeric_columns] )

In [149]:
# Train Logistic Regression
start_time = time.time()

model = LogisticRegression(max_iter=1000)

model.fit(
    X_train_label,
    y_train
)

training_time_label = time.time() - start_time

In [125]:
# Prediction
y_pred_label = model.predict(X_test_label)

# Accuracy
accuracy_label = accuracy_score(
    y_test,
    y_pred_label
)

print("Accuracy:", accuracy_label)
print("Training time:", training_time)
print("Feature count:", X_train_label.shape[1])

Accuracy: 0.8761164151524484
Training time: 0.2474982738494873
Feature count: 22


### Strategy 2: One-Hot Encoding

In [126]:
from sklearn.compose import ColumnTransformer

# Create fresh copy
X_train_onehot = X_train.copy()
X_test_onehot = X_test.copy()

In [127]:
# Find categorical columns
categorical_columns = X_train_onehot.select_dtypes(
    include=["str", "object"]
).columns


# Find numerical columns
numeric_columns = X_train_onehot.select_dtypes(
    include=["int64", "float64"]
).columns


In [128]:
# One Hot Encoder
onehot_encoder = OneHotEncoder(
    handle_unknown="ignore"
)

# Apply encoding only to categorical columns
preprocessor = ColumnTransformer(
    transformers=[
        (
            "cat",
            onehot_encoder,
            categorical_columns
        )
    ],
    remainder="passthrough"
)


# Fit on train, transform train/test
X_train_onehot = preprocessor.fit_transform(
    X_train_onehot
)

X_test_onehot = preprocessor.transform(
    X_test_onehot
)

In [129]:
# Applying Scaler 
scaler = StandardScaler(with_mean=False)

X_train_onehot = scaler.fit_transform(
    X_train_onehot
)

X_test_onehot = scaler.transform(
    X_test_onehot
)

In [130]:
# Train model
start_time = time.time()

model_onehot = LogisticRegression(
    max_iter=1000
)

model_onehot.fit(
    X_train_onehot,
    y_train
)

training_time_onehot = time.time() - start_time

In [131]:
# Prediction
y_pred_onehot = model_onehot.predict(
    X_test_onehot
)

accuracy_onehot = accuracy_score(
    y_test,
    y_pred_onehot
)

print("One-Hot Encoding")
print("Accuracy:", accuracy_onehot)
print("Training time:", training_time_onehot)
print("Feature count:", X_train_onehot.shape[1])

One-Hot Encoding
Accuracy: 0.8751924853711118
Training time: 0.35368800163269043
Feature count: 27


#### Strategy 3: Ordinal Encoding

In [132]:
# Fresh copy
X_train_ordinal = X_train.copy()
X_test_ordinal = X_test.copy()

In [133]:
# Find categorical columns
categorical_columns = X_train_ordinal.select_dtypes(
    include=["str", "object"]
).columns

In [134]:
# Ordinal encoder
ordinal_encoder = OrdinalEncoder(
    handle_unknown="use_encoded_value",
    unknown_value=-1
)

# Fit only on train data
X_train_ordinal[categorical_columns] = ordinal_encoder.fit_transform(
    X_train_ordinal[categorical_columns]
)

# Transform test using same encoder
X_test_ordinal[categorical_columns] = ordinal_encoder.transform(
    X_test_ordinal[categorical_columns]
)

In [135]:

# Scale numerical columns
scaler = StandardScaler()

numeric_columns = X_train_ordinal.select_dtypes( include=["int64", "float64"]).columns

X_train_ordinal[numeric_columns] = scaler.fit_transform(
    X_train_ordinal[numeric_columns]
)

X_test_ordinal[numeric_columns] = scaler.transform(
    X_test_ordinal[numeric_columns]
)

In [136]:
# Train model
start_time = time.time()

model_ordinal = LogisticRegression(
    max_iter=1000
)

model_ordinal.fit(
    X_train_ordinal,
    y_train
)

training_time_ordinal = time.time() - start_time

In [137]:
# Prediction
y_pred_ordinal = model_ordinal.predict(
    X_test_ordinal
)

accuracy_ordinal = accuracy_score(
    y_test,
    y_pred_ordinal
)

print("Ordinal Encoding")
print("Accuracy:", accuracy_ordinal)
print("Training time:", training_time_ordinal)
print("Feature count:", X_train_ordinal.shape[1])


Ordinal Encoding
Accuracy: 0.8761164151524484
Training time: 0.21422171592712402
Feature count: 22


### Comparision Table among 3 strategies:

In [153]:
results = pd.DataFrame({
    "Encoding": [
        "Label",
        "One-Hot",
        "Ordinal"
    ],
    "Accuracy": [
        accuracy_label,
        accuracy_onehot,
        accuracy_ordinal
    ],
    "Training Time": [
        training_time_label,
        training_time_onehot,
        training_time_ordinal
    ],
    "Feature Count" :[
        X_train_label.shape[1],
        X_train_onehot.shape[1],
        X_train_ordinal.shape[1]
    ]
})

results

,Encoding,Accuracy,Training Time,Feature Count
0,Label,0.876116,0.275539,22
1,One-Hot,0.875192,0.353688,27
2,Ordinal,0.876116,0.214222,22


I recommend using Ordinal Encoding because it gave the best overall result. It matched Label Encoding with the highest accuracy of 0.876116, trained the fastest at 0.214222 seconds, and used only 22 features, while One-Hot Encoding was slower and slightly less accurate.